casper/src/main/resources/Registry.rho

```diff
@@ -330,104 +358,241 @@ in {
+              ops!("buildUri", pubKeyBytes, *uriCh) |
+              for (@uri <- uriCh) {
+                for (@map <<- _registryStore) {
+                  new responseCh in {
+                    TreeHashMap!("get", map, uri, *responseCh) |
+                    for (@response <- responseCh) {
+                      match response {
+                        Nil => {
+                          new ack in {
+                            TreeHashMap!("set", map, uri, (nonce, data), *ack) |
+                            for (_ <- ack) {
+                              ret!(uri)
+                            }
+                          }
+                        }
+                        (oldNonce, _) => {
+                          if (nonce > oldNonce) {
+                            new ack in {
+                              TreeHashMap!("set", map, uri, (nonce, data), *ack) |
+                              for (_ <- ack) {
+                                ret!(uri)
+                              }
+                            }
+                          } else {
+                            ret!(Nil)
+                          }
                         }
                       }
                     }
-                    (oldNonce, _) => {
-                      if (nonce > oldNonce) {
-                        new ack in {
-                          TreeHashMap!("set", map, uri, (nonce, data), *ack) |
-                          for (_ <- ack) {
-                            ret!(uri)
+                  }
+                }
+              }
+            }
+          }
+        }
+      } |
+    {
+      // ------
+      // Update
+      // deployerID: (`rho:rchain:deployerId`)
+      // uri:   the uri you want to update.  `rho:rchain:pos` etc.
+      // con:   the update consume which take the old state
+      // meth:  the target new contract which would replace the old one
+      // ret:   channel which returns the update result
+      contract systemContractManager(@"update", @securityData, @uri, @con, ret) = {
+        // there is an update proposal for `uri`
+        new rl(`rho:registry:lookup`), rlRet, locRet in {
+          rl!(uri, *rlRet)|
+          for (@(_, dispatcher) <- rlRet) {
+            @(*locToken, dispatcher)!(*locRet)|
+            for (loc <- locRet) {
+              for (oldLib <<- loc){
+                new securityCh in {
+                  @(*oldLib, "securityCheck")!(securityData, *securityCh)|
+                  for (@res <-securityCh){
+                    match res {
+                      (true, Nil) => {
+                        for (oldLib <- loc) {
+                          new tmp, newRet in {
+                            @(*oldLib, "extractState")!(*tmp) |
+                            for (@oldState <- tmp) {
+                              // launch new contract instance with initial state extracted from the old instance
+                              @con!(oldState, *oldLib, *newRet) |
```

Comment:
So here owner of the original (oldLib) contract will receive the instance. It means it can call `@(*oldLib, "extractState")` if needed so there is no need to do it as part of _systemContractManager_ code.
We can do the same for security check.

Reply:
User is providing `@con` consume which will receive old lib. What I pointing here is that user can call `@(*oldLib, "extractState")` if needed so it doesn't have to be supplied together with old lib. This is of course after security check is done as you already done.

Reply:
> We can do the same for security check.

I meant this part. How could we do that the same for security check.

Reply:
You are right, we cannot do this for security check. Before passing security check old lib cannot be accessed by the caller.